fluent/client/ws/connection.go

```diff
@@ -0,0 +1,197 @@
+package ws
+
+import (
+	"errors"
+	"io"
+	"log"
+	"runtime/debug"
+	"sync"
+	"time"
+
+	ext "github.com/IBM/fluent-forward-go/fluent/client/ws/ext"
+	"github.com/gorilla/websocket"
+)
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -generate
+
+type ReadHandler func(conn Connection, messageType int, p []byte, err error) error
+
+type ConnectionOptions struct {
+	CloseHandler func(conn Connection, code int, text string) error
+	PingHandler  func(conn Connection, appData string) error
+	PongHandler  func(conn Connection, appData string) error
+	ReadDeadline time.Time
+	ReadHandler  ReadHandler
+	// A zero value for means writes will not time out.
+	WriteDeadline time.Time
+}
+
+//counterfeiter:generate . Connection
+type Connection interface {
+	ext.Conn
+	ReadHandler() ReadHandler
+	Listen() error
+	SetReadHandler(rh ReadHandler)
+	CloseWithMsg(closeCode int, msg string) error
+	Write(data []byte) (int, error)
+}
+
+type connection struct {
+	ext.Conn
+	errChan      chan error
+	writeLock    sync.Mutex
+	closedLock   sync.Mutex
+	readHandler  ReadHandler
+	connClosed   bool
+	closeErrChan sync.Once
+}
+
+func NewConnection(conn ext.Conn, opts ConnectionOptions) Connection {
+	wsc := &connection{
+		Conn:    conn,
+		errChan: make(chan error),
+	}
+
+	if opts.CloseHandler != nil {
+		wsc.SetCloseHandler(func(code int, text string) error {
+			return opts.CloseHandler(wsc, code, text)
+		})
+	}
+
+	if opts.PingHandler != nil {
+		wsc.SetPingHandler(func(appData string) error {
+			return opts.PingHandler(wsc, appData)
+		})
+	}
+
+	if opts.PongHandler != nil {
+		wsc.SetPongHandler(func(appData string) error {
+			return opts.PongHandler(wsc, appData)
+		})
+	}
+
+	if opts.ReadHandler != nil {
+		wsc.SetReadHandler(opts.ReadHandler)
+	}
+
+	wsc.SetReadDeadline(opts.ReadDeadline)
+	wsc.SetWriteDeadline(opts.WriteDeadline)
```

Comment:
errors are returned but not handled for SetReadDeadline and SetWriteDeadline. Do the opts passed need to be validated?